bayes_kit/ensemble.py

```diff
@@ -1,68 +1,91 @@
+        proposal bounds, number of walkers and initial parameter values.
+            
+        Parameters:
+        model: class used to evaluate log densities
+        a: bounds on proposal (default 2)
+        walkers: an even number of walkers to use (default dimensionality of `model * 2`)
+        init: `walker` x `dimensio`n array of initial positions (defaults to standard normal)
+
+        Throws:
+        ValueError: if `a` is provided and not >= 1, `walker`s is provided and not strictly positive and even,
+        or if the `init` is provided and is not an `NDArray` of shape `walker` x `dimension`
+        """
+        self._model = model
+        self._dim = self._model.dims()
+        if a != None and np.float64(a) < 1:
+            raise ValueError(f"stretch bound must be greater than or equal to 1; found {a=}")
+        self._a = np.float64(a or 2.0)
+        self._sqrt_a = np.sqrt(np.float64(a))
+        self._inv_sqrt_a = 1 / self._sqrt_a
+        self._walkers = np.int64(walkers or 2 * self._dim)
+        if self._walkers < 2 or self._walkers % 2 != 0:
+            raise ValueError(f"walkers must be strictly positive, even integer; found {walkers=}")
+        self._halfwalkers = self._walkers // 2
+        self._drawshape = (int(self._walkers), self._dim)
+        self._thetas = np.asarray(init or np.random.normal(size=self._drawshape))
+        if self._thetas.shape != self._drawshape:
+            raise ValueError(f"init must be shape of draw {self._drawshape}; found {self._thetas.shape=}")
+        self._firsthalf = range(0, int(self._halfwalkers))
+        self._secondhalf = range(int(self._halfwalkers), int(self._walkers))
 
-    # def __iter__(self):
-    #     return self
+    def __iter__(self) -> Iterator[Sample]:
+        return self
 
-    # def __next__(self):
-    #     return self.sample
+    def __next__(self) -> Sample:
+        return self.sample()
 
-    # def draw_z(self):
-    #     """Return random draw z in (1/a, a) with p(z) propto 1 / sqrt(z)"""
-    #     return np.square(np.random.uniform(self._inv_sqrt_a, self._sqrt_a))
+    def draw_z(self) -> Sample:
+        """Return random draw z in (1/a, a) with p(z) propto 1 / sqrt(z)"""
+        return np.asarray(np.square(np.random.uniform(self._inv_sqrt_a, self._sqrt_a)))
 
-    # def stretch_move(self, theta_k: NDarray[np.float64], theta_j: NDarray[np.float64]):
-    #     z = self.draw_z()
-    #     theta_star = theta_j + z * (theta_k - theta_j)  # (1 - z) * theta_j + z * theta_k
-    #     log_q = (self._dims - 1) * np.log(z) + self._model.log_density(theta_star) - self._model.log_density(theta_k)
-    #     if np.log(np.random.uniform()) < log_q:
-    #         return theta_star
-    #     return theta_k
+    def stretch_move(self, theta_k: NDArray[np.float64], theta_j: NDArray[np.float64]) -> Sample:
+        z = self.draw_z()
+        theta_star = np.asarray(theta_j + z * (theta_k - theta_j))  # (1 - z) * theta_j + z * theta_k
```

Comment:
Same note on np.asarray. In general we should almost never be calling np.asarray/asanyarray except for if the object is user-provided. If we're working around limitations in numpy's type hints, we should either use an explicit declaration like `theta_star: NDArray[np.float64]` or use a `# type: ignore` comment

Reply:
OK.  I'm not sure how to figure out if I'm actually getting the right types with a `# type: ignore` nor am I sure what the performance hit is for declaring `theta_star: NDArray[np.float64]`.  When will typing like this impose a runtime performance penalty?

Reply:
The only real way to be sure when using `type: ignore` is to put a runtime assertion or read (and trust!) the documentation of the function. It is generally safe to assume a numpy function will return an ndarray when you expect one, so this isn’t too controversial.

If the thing following the colon is “simple” (in general it can be any python expression which resolves to a type, which obviously could be arbitrarily expensive), like you have there, then that form has basically no cost other than the fact that it is parsed.

Reply:
I managed to get rid of the `asarray` functions and `# type: ignore` directives by explicitly typing LHS variables.